apollo-router/src/spec/query/change.rs

```diff
@@ -132,158 +214,268 @@ impl<'a> QueryHashVisitor<'a> {
+        self.hashed_types.insert(name.to_string());
 
-        if let Some(ty) = self.schema.types.get(t) {
+        if let Some(ty) = self.schema.types.get(name) {
             self.hash_extended_type(ty)?;
         }
+        "^TYPE_BY_NAME-END".hash(self);
+
         Ok(())
     }
 
     fn hash_extended_type(&mut self, t: &'a ExtendedType) -> Result<(), BoxError> {
+        "^EXTENDED_TYPE".hash(self);
+
         match t {
             ExtendedType::Scalar(s) => {
+                "^SCALAR".hash(self);
+
+                "^DIRECTIVE_LIST".hash(self);
                 for directive in &s.directives {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIST_END".hash(self);
             }
+            // this only hashes the type level info, not the fields, because those will be taken from the query
+            // we will still hash the fields using for the key
             ExtendedType::Object(o) => {
+                "^OBJECT".hash(self);
+
+                "^DIRECTIVE_LIST".hash(self);
                 for directive in &o.directives {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIST_END".hash(self);
 
                 self.hash_join_type(&o.name, &o.directives)?;
+                "^OBJECT_END".hash(self);
+
+                "^IMPLEMENTED_INTERFACES_LIST".hash(self);
+                for interface in &o.implements_interfaces {
+                    self.hash_type_by_name(&interface.name)?;
+                }
+                "^IMPLEMENTED_INTERFACES_LIST_END".hash(self);
             }
             ExtendedType::Interface(i) => {
+                "^INTERFACE".hash(self);
+
+                "^DIRECTIVE_LIST".hash(self);
                 for directive in &i.directives {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIST_END".hash(self);
+
                 self.hash_join_type(&i.name, &i.directives)?;
+
+                "^IMPLEMENTED_INTERFACES_LIST".hash(self);
+                for implementor in &i.implements_interfaces {
+                    self.hash_type_by_name(&implementor.name)?;
+                }
+                "^IMPLEMENTED_INTERFACES_LIST_END".hash(self);
+
+                if let Some(implementers) = self.schema().implementers_map().get(&i.name) {
+                    "^IMPLEMENTER_OBJECT_LIST".hash(self);
+
+                    for object in &implementers.objects {
+                        self.hash_type_by_name(object)?;
+                    }
+                    "^IMPLEMENTER_OBJECT_LIST_END".hash(self);
+
+                    "^IMPLEMENTER_INTERFACE_LIST".hash(self);
+                    for interface in &implementers.interfaces {
+                        self.hash_type_by_name(interface)?;
+                    }
+                    "^IMPLEMENTER_INTERFACE_LIST_END".hash(self);
+                }
+
+                "^INTERFACE_END".hash(self);
```

Comment:
Interface has end, but other types doesn't
For example there are no `^OBJECT_END` or `^UNION_END`
I don't think it's exploitable but I think we need consistency here, so better to add end "keywords" everywhere